scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+)
+```
+notes:
+Then we'll apply a map function that multiplies the number by 2.
+
+---
+```ts [6-7]
+const program = pipe(
+  getNumber(),
+  // Effect<never, never, number>
+  Effect.map((x) => x * 2),
+  // Effect<never, never, number>
+  Effect.flatMap(checkIfAtLeastFive),
+  // Effect<never, Error, number>
+)
+```
+notes:
+Then we'll flatMap with the check if at least 5 function. Notice that the value of our program now contains the error from that function.
+
+---
+```ts [8-9]
+const program = pipe(
+  getNumber(),
+  // Effect<never, never, number>
+  Effect.map((x) => x * 2),
+  // Effect<never, never, number>
+  Effect.flatMap(checkIfAtLeastFive),
+  // Effect<never, Error, number>
+  Effect.flatMap(logNumber)
+  // Effect<never, Error, void>
+);
+```
+
+notes:
+finally we'll flat map with our log function. The final program is of type Effect never Error void, meaning it either fails with type Error or succeeds with type void.
+
+---
+
+the value was consumed by logNumber
+
+```ts
+pipe(
+	Effect.succeed(5),
+	Effect.flatMap(logNumber),
+	Effect.map(x => x + 1)
+	// ^ ERROR: 
+	// '+' cannot be applied to types 'void' and 'number'.
+)
+```
+
+notes:
+What if we want to cause some side effect, but still have access to the original value after?
+
+---
+
+Execute side effects with tap
+```ts
+const program = pipe(
+	Effect.succeed(5),
+	Effect.tap(logNumber),
+	Effect.map(x => x + 1)
+)
+
+Effect.runSync(program); // 6
+// console: 5
+```
+
+notes:
+We can do this with the tap function. Tap is just like flatMap but the result of the function passed to is it ignored, and the original value is still available for the next function. This is useful for executing side effects without altering the result of an Effect
+
+---
+
+Consuming multiple Effects with all
+
+```ts
+const foo = Effect.succeed(42)
+const bar = Effect.succeed("Hello")
+ 
+// type: Effect<never, never, [number, string]>
+const combinedEffect = Effect.all([foo, bar])
+ 
+console.log(Effect.runSync(combinedEffect))
+// console: [42, "Hello"]
+```
+
+notes:
+So far we've only been working with single values, but what about transforming the results of multiple Effects.
+
+For this we can use all. All takes an array of effects and returns an Effect whose value is a tuple of the value types of the passed effects.
+
+Here in this example, we merge foo and bar to a single effect then run and log it.
+
+---
+
+A real program!
+
+notes:
+And thats all we need to get started on the example program we will make and improve over the rest of the video.
+
+---
+
+Find the heaviest pokemon!
+
+notes:
+Our program will look like this:
+Generate some random numbers
+using the numbers as ids, get information about the pokemon with that id from an api
+validate the returned json
+find the heaviest pokemon and log it
+
+It will simple but naïve and we'll add robustness bit by bit.
+
+---
+```ts
+// type: Effect<never, unknown, any>
+const getPokemon = (id: number) =>
```

Comment:
I'd use the variant that allows for mapping, removing `unknown` from the error type. Note that `unknown` is destructive, as soon as you mix `unknown | X` the result would be `unknown` regardless of `X`